src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,180 @@
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+    }
+
+    @Override
+    public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
+    }
+
+    @Override
+    public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            final Node<T> newNode = new Node<>(current.prev, value, current);
+            if (current.prev == null) {
+                head = newNode;
+            } else {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+            size++;
+        }
+    }
+
+    @Override
+    public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
+    }
+
+    @Override
+    public T get(int index) {
+        checkIndex(index);
+        return getNodeByIndex(index).value;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index < (size >> 1)) {
+            Node<T> node = head;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        } else {
+            Node<T> node = tail;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+            return node;
+        }
+    }
+
+    @Override
+    public T set(T value, int index) {
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        T finalValue = current.value;
+        current.value = value;
+        return finalValue;
+    }
+
+    @Override
+    public T remove(int index) {
+        checkIndex(index);
+        if (index == 0) {
+            final T temp = (T) head.value;
+            head = head.next;
+            size--;
+            if (head == null) {
+                tail = null;
+            }
+            return temp;
+        } else if (index == size - 1) {
+            if (size == 0) {
+                return null;
+            } else {
+                final T temp = (T) tail.value;
+                tail = tail.prev;
+                tail.next = null;
+                size--;
+                return temp;
+            }
+        } else {
+            Node<T> previous = head;
+            for (int i = 0; i < index; i++) {
+                previous = previous.next;
+            }
+            Node<T> current = previous.next;
+            previous.next = current;
+            current.prev = previous.prev;
+            previous.prev.next = current;
+            size--;
+            return previous.value;
+        }

```

Comment:
Should be something like below. Recheck if getNodeByIndex() and unlink() are ready for such usage but they should be
```suggestion
        Node<T> nodeToRemove =  getNodeByIndex(index);
        T result = nodeToRemove.value;
        unlink(nodeToRemove);
        return result;
```